File: src/main/resources/static/js/entityManager.js
/*

entityManager.js

A module which handles rendering of entities

*/


"use strict";

/*jslint nomen: true, white: true, plusplus: true*/


let entityManager = {

// "PRIVATE" DATA
_gameObjects   : [],
posShift : {x : 0, y : 0},
_removedIDs : [],
_playerID : 0,
_playerPos : {x : 0, y : 0},
_stats : {health : 3, drink: 3, breath : 10, hasTreasure : false, money: 0, hasMap : false, markerRot : 0},
_background : null,
changedWorld : true,
isInShop : false,
hasGeneratedShopUI: false,

// "PUBLIC" METHODS

deferredSetup : function () {

},


render: function(ctx) {
    if(this._background != null)
        this._background.render(ctx);
    else{
        this._background = new GameObject({
            id : -1,
            targetX : 0,
            targetY : 0,
            sprite : 'background',
            scaleX: 2,
            scaleY: 2,
            isStatic : true,
            zIndex : -1,
            tooltip: "",
            isRendered: true
        })
    }

    //Map to temporary array
    let map = this._gameObjects.map(function (el, index) {
        return { index : index, value : el.zIndex };
    });

    //Sort by z index
    map.sort(function (a, b) {
        return a.value - b.value;
    });

    //Rebuild sorted array
    let objectsSorted = map.map(function (el) {
        return entityManager._gameObjects[el.index];
    });

    //Render objects
    for (let i = 0; i < objectsSorted.length; i++) {
        objectsSorted[i].render(ctx)
    }

    renderUI(this._stats.health, this._stats.drink, this._stats.breath, this._stats.hasTreasure,
            this._stats.hasMap, this._stats.markerRot);
},

updateGameState: function(response) {
    if(response == null)
        return;

    let gameState = JSON.parse(response);

    if(gameState.shopItems != null){
        this.isInShop = true;
        if(!this.hasGeneratedShopUI){
            renderShopUI(gameState.shopItems);
            this.hasGeneratedShopUI = true;
        }
        $("#money").html("GOLD: " + this._stats.money);
    } else{
        if(this.hasGeneratedShopUI)
            clearShopUI();
    }

    if(gameState.playerID != null){
        this._playerID = gameState.playerID
    }

    if(gameState.stats != null) {
        this._stats.health = gameState.stats.health;
        this._stats.drink = gameState.stats.drink;
        this._stats.breath = gameState.stats.breath;
        this._stats.hasTreasure = gameState.stats.hasTreasure;
        this._stats.money = gameState.stats.money;
        this._stats.hasMap = gameState.stats.hasMap;
        this._stats.markerRot = gameState.stats.markerRot;
    }


    gameState.gameObjects.forEach(go => {
        if(go.id === this._playerID){
            this._playerPos.x = go.x;
            this._playerPos.y = go.y;
        }
        if(this._gameObjects.filter(e => e.id === go.id ).length > 0){
            let foundIndex = this._gameObjects.findIndex(e => e.id === go.id);
            this._gameObjects[foundIndex].targetX = go.x;
            this._gameObjects[foundIndex].targetY = go.y;
            this._gameObjects[foundIndex].scaleX = go.scaleX;
            this._gameObjects[foundIndex].scaleY = go.scaleY;
            this._gameObjects[foundIndex].zIndex = go.zIndex;
            this._gameObjects[foundIndex].isRendered = go.isRendered;
            if(this._playerID === parseInt(go.id))
                this._gameObjects[foundIndex].tooltip = go.tooltip;
        } else {
            let obj = new GameObject({
                id : go.id,
                targetX : go.x,
                targetY : go.y,
                sprite : go.sprite,
                scaleX: go.scaleX,
                scaleY: go.scaleY,
                isStatic : go.isStatic,
                zIndex : go.zIndex,
                tooltip: "",
                isRendered: go.isRendered
            });
            this._gameObjects.push(obj);
        }
    });

    gameState.tempRemovedGameObjectIDs.forEach(ID => {
        let index = this._gameObjects.findIndex(e => e.id === ID);
        this._gameObjects.splice(index, 1);
    });

    gameState.removedGameObjectIDs.forEach(ID => {
        if(!this._removedIDs[ID]) {
            let index = this._gameObjects.findIndex(e => e.id === ID);
            this._removedIDs[ID] = true;
            this._gameObjects.splice(index, 1);
        }
    });

    if(this._background != null) {
        this._background.targetX = -this.posShift.x;
        this._background.targetY = -this.posShift.y;
    }

    if(gameState.changedWorld != null){
        this.changedWorld = gameState.changedWorld;
    }

    if(!this.changedWorld) {
        let xLerp = util.lerp(0, gameState.posShift.x - this.posShift.x, 0.03);
        let yLerp = util.lerp(0, gameState.posShift.y - this.posShift.y, 0.03);
        g_ctx.translate(xLerp, yLerp);

        this.posShift.x += xLerp;
        this.posShift.y += yLerp;
    } else{
        g_ctx.translate(gameState.posShift.x - this.posShift.x, gameState.posShift.y - this.posShift.y);
        this.posShift.x += gameState.posShift.x - this.posShift.x;
        this.posShift.y += gameState.posShift.y - this.posShift.y;
    }

}

};

// Some deferred setup which needs the object to have been created first
entityManager.deferredSetup();

File: AndroidProject/PirateGame/app/src/main/java/com/example/pirategame/EntityManager.java
package com.example.pirategame;

import android.app.Activity;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.util.Pair;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;

public class EntityManager {

    private Map<Integer, GameObject> gameObjects = new LinkedHashMap<>();
    private Map<Integer, GameObject> backgroundObjects = new LinkedHashMap<>();

    private Queue<Pair<Integer, GameObject>> backgroundObjectQueue = new LinkedList<>();

    private Pair<Float, Float> posShift = new Pair<>(0.0f, 0.0f);
    int[] removedIDs;
    private int playerID;
    private Pair<Float, Float> playerPos = new Pair<>(0.0f, 0.0f);
    private Stats stats = new Stats();
    private GameObject background;
    boolean changedWorld = true;
    private boolean isInShop;
    private boolean hasGeneratedShopUI;

    public boolean isLoggedIn;
    private boolean gameObjectsMarkedDirty;

    public EntityManager(){
        Start();
    }


    private void Start(){
    }

    public void updateGameState(JSONObject gameState, Canvas ctx) throws JSONException {
        {
            //TODO: Implement shop UI
            if(gameState.has("shopItems")){
                this.isInShop = true;
                if(!this.hasGeneratedShopUI){
                    //renderShopUI(gameState.shopItems);
                    this.hasGeneratedShopUI = true;
                }
                //$("#money").html("GOLD: " + this._stats.money);
            } else{
                //if(this.hasGeneratedShopUI)
                    //clearShopUI();
            }



            if(gameState.has("playerID")){
                this.playerID = gameState.getInt("playerID");
            }

            if(gameState.has("stats")) {
                JSONObject stats = (JSONObject) gameState.get("stats");
                this.stats.setHealth(stats.getInt("health"));
                this.stats.setDrink(stats.getInt("drink"));
                this.stats.setBreath(stats.getInt("breath"));
                this.stats.setHasTreasure(stats.getBoolean("hasTreasure"));
                this.stats.setMoney(stats.getInt("money"));
                this.stats.setHasMap(stats.getBoolean("hasMap"));
                this.stats.setMarkerRot(stats.getDouble("health"));
            }

            JSONArray gameObjects = gameState.getJSONArray("gameObjects");

            //Iterate through all received game objects
            for(int i = 0; i < gameObjects.length(); i++){
                JSONObject go = (JSONObject) gameObjects.get(i);
                if(go.getInt("id") == this.playerID)
                    this.playerPos= new Pair<>((float)go.getDouble("x"),
                                               (float)go.getDouble("y"));

                int foundID = go.getInt("id");
                boolean hasFoundObject = false;

                //Check if game object exists
                if(this.gameObjects.containsKey(foundID) || this.backgroundObjects.containsKey(foundID)){
                    hasFoundObject = true;
                    GameObject obj = this.gameObjects.get(foundID);
                    if(obj == null) continue;
                    obj.targetX = (float) go.getDouble("x");
                    obj.targetY = (float) go.getDouble("y");
                    obj.scaleX = (float) go.getDouble("scaleX");
                    obj.scaleY = (float) go.getDouble("scaleY");
                    obj.zIndex = go.getInt("zIndex");
                    obj.isRendered = go.getBoolean("isRendered");
                    if (this.playerID == foundID)
                        obj.tooltip = go.getString("tooltip");
                }

                //If no instance exists, create a new instance
                if(!hasFoundObject) {
                    GameObject obj = new GameObject(
                            go.getInt("id"),
                            (float) go.getDouble("x"),
                            (float) go.getDouble("y"),
                            (float) go.getDouble("scaleX"),
                            (float) go.getDouble("scaleY"),
                            go.getInt("zIndex"),
                            go.getString("sprite"),
                            "",
                            go.getBoolean("isStatic"),
                            go.getBoolean("isRendered"),
                            this
                    );
                    if(obj.zIndex == 0) //If background object
                        this.backgroundObjectQueue.add(new Pair<>(obj.id, obj));
                    else
                        this.gameObjects.put(obj.id, obj);
                    gameObjectsMarkedDirty = true;
                }

            }

            JSONArray tempRemovedGameObjects = gameState.getJSONArray("tempRemovedGameObjectIDs");
            for(int i = 0; i < tempRemovedGameObjects.length(); i++) {
                int ID = ((JSONObject) tempRemovedGameObjects.get(i)).getInt("id");
                int index = 0;
                for(int k = 0; k < this.gameObjects.size(); k++){
                    if(this.gameObjects.get(k).id == ID) {
                        index = k;
                        break;
                    }
                }
                this.gameObjects.remove(index);
            }

            JSONArray removedGameObjects = gameState.getJSONArray("removedGameObjectIDs");
            for(int i = 0; i < removedGameObjects.length(); i++) {
                int ID = (removedGameObjects.getInt(i));
                int index = 0;
                for(int k = 0; k < this.gameObjects.size(); k++){
                    if(this.gameObjects.get(k).id == ID) {
                        index = k;
                        break;
                    }
                }
                this.gameObjects.remove(index);

            }

            if(this.background != null) {
                this.background.targetX = -this.posShift.first;
                this.background.targetY = -this.posShift.second;
            }

            if(gameState.has("changedWorld")){
                this.changedWorld = gameState.getBoolean("changedWorld");
            }

            JSONObject posShift = (JSONObject) gameState.get("posShift");

            float x = (float) posShift.getDouble("x");
            float y = (float) posShift.getDouble("y");
            if(!this.changedWorld) {
                float xLerp = Util.lerp(0, (x-800) - this.posShift.first, 0.03f);
                float yLerp = Util.lerp(0, (y-300) - this.posShift.second, 0.03f);

                //System.out.println("X: " + xLerp + "Y: " + yLerp);

                this.posShift = new Pair<>(this.posShift.first + xLerp, this.posShift.second + yLerp);
            } else{
                float xLerp = Util.lerp(0, x - this.posShift.first, 0.03f);
                float yLerp = Util.lerp(0, y - this.posShift.second, 0.03f);

                this.posShift = new Pair<>(this.posShift.first + xLerp, this.posShift.second + yLerp);
            }
        }
    }

    public void render(Canvas ctx, boolean isBackground, GameView gameView){
        if(!isBackground) {
            if(GameActivity.gButton == null){
                GameActivity.gButton = new GButton(100, 100,
                        BitmapFactory.decodeResource(gameView.getResources(),
                                Util.StringToBitmap("map")));
            }

            GameActivity.gButton.setPosition(playerPos.first + 500, playerPos.second + 500);
            GameActivity.gButton.x = playerPos.first + 500;
            GameActivity.gButton.y = playerPos.second + 500;
            GameActivity.gButton.draw(ctx);

            ctx.translate(this.posShift.first, this.posShift.second);

            if (gameObjects.size() > 0) {
                //Sort gameObjects via zIndex
                if (gameObjectsMarkedDirty) {
                    gameObjects = sortByValue(gameObjects);
                    gameObjectsMarkedDirty = false;
                }

                gameObjects.forEach((id, obj) -> {
                    obj.render(ctx, gameView);
                });
            }
        } else {
            ctx.translate(posShift.first, posShift.second);

            if(background != null && !background.hasBeenRendered){
                background.render(ctx, gameView);
                background.hasBeenRendered = true;
            } else{
                background = new GameObject(-1, 0, 0,
                        2, 2, -1,
                        "background", "",
                        true, true,
                        this);
            }

            while(!backgroundObjectQueue.isEmpty()){
                Pair<Integer, GameObject> p = backgroundObjectQueue.peek();
                if(p != null) {
                    backgroundObjects.put(p.first, p.second);
                    backgroundObjectQueue.remove();
                }
            }

            backgroundObjects.forEach((id, obj) -> {
                //if (!obj.hasBeenRendered) {
                    obj.render(ctx, gameView);
                    obj.hasBeenRendered = true;
                //}
            });

        }
    }

    private static <K, V extends Comparable<? super V>> Map<K, V> sortByValue(Map<K, V> map) {
        List<Map.Entry<K, V>> list = new ArrayList<>(map.entrySet());
        list.sort(Map.Entry.comparingByValue());

        Map<K, V> result = new LinkedHashMap<>();
        for (Map.Entry<K, V> entry : list) {
            result.put(entry.getKey(), entry.getValue());
        }

        return result;
    }

    public Pair<Float, Float> getPosShift() {
        return posShift;
    }
}
